File: src/main/java/net/azisaba/bungeeproxyannouncer/PlayerIPAddressList.java
package net.azisaba.bungeeproxyannouncer;

import io.netty.handler.codec.haproxy.HAProxyMessage;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.util.AbstractMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class PlayerIPAddressList {
    public static final Map<Map.Entry<String, Integer>, String> map = new ConcurrentHashMap<>();

    @SuppressWarnings("unused") // used
    public static void handle(HAProxyMessage message, SocketAddress proxyHost) {
        if (!(proxyHost instanceof InetSocketAddress address)) return;
        map.put(new AbstractMap.SimpleImmutableEntry<>(message.sourceAddress(), message.sourcePort()), address.getAddress().getHostAddress().replaceFirst("(.*)%.*", "$1"));
    }

    public static void remove(SocketAddress playerHost) {
        if (!(playerHost instanceof InetSocketAddress address)) return;
        map.remove(new AbstractMap.SimpleImmutableEntry<>(address.getAddress().getHostAddress(), address.getPort()));
    }
}

File: src/main/kotlin/net/azisaba/bungeeProxyAnnouncer/PlayerIPAddressList.kt
package net.azisaba.bungeeProxyAnnouncer

import io.netty.handler.codec.haproxy.HAProxyMessage
import java.net.InetSocketAddress
import java.net.SocketAddress

object PlayerIPAddressList {
    val map = mutableMapOf<Pair<String, Int>, String>()

    @Suppress("unused")
    @JvmStatic
    fun handle(message: HAProxyMessage, proxyHost: SocketAddress) {
        if (proxyHost !is InetSocketAddress) return
        message.sourceAddress()
        map[message.sourceAddress() to message.sourcePort()] = proxyHost.address.hostAddress.replaceFirst("(.*)%.*".toRegex(), "$1")
    }

    @Suppress("unused")
    @JvmStatic
    fun remove(playerHost: SocketAddress) {
        if (playerHost !is InetSocketAddress) return
        map.remove(playerHost.address.hostAddress to playerHost.port)
    }
}
